File: test/language/templates/cpp98/ExpressionConstAfterNestedTemplateArgumentTest.cpp
#include "gtest/gtest.h"

#include "templates/expression_const_after_nested_template_argument/ConstAfterNested.h"

namespace templates
{
namespace expression_const_after_nested_template_argument
{

TEST(ExpressionConstAfterNestedTemplateArgumentTest, readWrite)
{

    ConstAfterNested constAfterNested;
    zserio::ObjectArray<Element_uint32>& array = constAfterNested.getCompound().getArray();
    array.resize(3);
    array[0].setValue(1);
    array[1].setValue(2);
    array[2].setValue(3);

    zserio::BitStreamWriter writer;
    constAfterNested.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    const ConstAfterNested readConstAfterNested(reader);

    ASSERT_TRUE(constAfterNested == readConstAfterNested);
}

} // namespace expression_const_after_nested_template_argument
} // namespace templates

File: test/language/templates/java/templates/ExpressionConstAfterNestedTemplateArgumentTest.java
package templates;

import static org.junit.Assert.*;

import java.io.File;
import java.io.IOException;

import org.junit.Test;

import zserio.runtime.BitSizeOfCalculator;
import zserio.runtime.ZserioError;
import zserio.runtime.array.ObjectArray;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.BitStreamWriter;
import zserio.runtime.io.FileBitStreamReader;
import zserio.runtime.io.FileBitStreamWriter;

import templates.expression_const_after_nested_template_argument.ConstAfterNested;
import templates.expression_const_after_nested_template_argument.Compound_Element_uint32_SIZE;
import templates.expression_const_after_nested_template_argument.Element_uint32;

public class ExpressionConstAfterNestedTemplateArgumentTest
{
    @Test
    public void readWrite() throws IOException
    {
        ObjectArray<Element_uint32> array = new ObjectArray<Element_uint32>(3);
        array.setElementAt(new Element_uint32(1), 0);
        array.setElementAt(new Element_uint32(2), 1);
        array.setElementAt(new Element_uint32(3), 2);

        final ConstAfterNested constAfterNested = new ConstAfterNested(new Compound_Element_uint32_SIZE(array));
        final BitStreamWriter writer = new FileBitStreamWriter(TEST_FILE);
        constAfterNested.write(writer);
        writer.close();
        final BitStreamReader reader = new FileBitStreamReader(TEST_FILE);

        final ConstAfterNested readConstAfterNested =
                new ConstAfterNested(reader);
        reader.close();
        assertTrue(constAfterNested.equals(readConstAfterNested));
    }

    private static final File TEST_FILE = new File("test.bin");
}
